packages/test/src/util/browser-tests-runner/browser-dependencies/index.js

```diff
@@ -0,0 +1,141 @@
+// Capture client logs and forward to server.
+["log", "info", "error"].forEach(method => {
+  const fn = console[method] || console.log || (() => {});
+  console[method] = (...args) => {
+    send(["console", method, args]);
+    fn.apply(console, args);
+  };
+});
+
+// Forward uncaught excpections.
+window.addEventListener("error", ({ error }) => console.error(error));
+
+require("chai").config.includeStack = true;
+const BrowserContext = require("./browser-context");
+let buffer = [];
+let options;
+
+try {
+  options = JSON.parse(decodeURIComponent(window.location.search.slice(1)));
+} catch (_) {
+  options = {};
+}
+
+// Apply mocha options.
+const mochaOptions = {
+  ...options.mochaOptions,
+  reporter: "spec",
+  useColors: true,
+  ui: "bdd"
+};
+
+Object.keys(mochaOptions).forEach(key => {
+  if (typeof mocha[key] === "function") {
+    mocha[key](mochaOptions[key]);
+  }
+});
+
+window.__run_tests__ = done => {
+  const socket = require("engine.io-client")(`ws://${location.host}`);
+  socket.once("open", () => {
+    const runner = mocha.run();
+    const fails = [];
+    let timeout;
+    queueFlush(); // We write the buffered console output on an interval.
+
+    runner.on("fail", (test, err) => {
+      fails.push({
+        name: test.title,
+        result: false,
+        message: err.message,
+        stack: err.stack,
+        titles: flattenTitles(test)
+      });
+    });
+
+    runner.on("end", () => {
+      window.mochaResults = runner.stats;
+      window.mochaResults.reports = fails;
+      clearTimeout(timeout);
+      flush(() =>
+        done({
+          success: !fails.length,
+          coverage: window.__coverage__
+        })
+      );
+    });
+
+    function flush(cb) {
```

Comment:
Is it a problem that you're never calling the callback here?

Reply:
Done gets called after flush unless I’m missing something 😜.